src/data-model-providers/codedriven/CodeDrivenDataModelProvider.cpp

```diff
@@ -0,0 +1,303 @@
+    {
+        for (const auto & path : cluster->GetPaths())
+        {
+            if (path.mEndpointId == endpointId)
+            {
+                ReturnErrorOnFailure(
+                    out.Append({ path.mClusterId, cluster->GetDataVersion(path), cluster->GetClusterFlags(path) }));
+            }
+        }
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CodeDrivenDataModelProvider::GeneratedCommands(const ConcreteClusterPath & path, ReadOnlyBufferBuilder<CommandId> & out)
+{
+    ServerClusterInterface * serverCluster = GetServerClusterInterface(path);
+    VerifyOrReturnError(serverCluster != nullptr, CHIP_ERROR_NOT_FOUND);
+    return serverCluster->GeneratedCommands(path, out);
+}
+CHIP_ERROR CodeDrivenDataModelProvider::AcceptedCommands(const ConcreteClusterPath & path,
+                                                         ReadOnlyBufferBuilder<DataModel::AcceptedCommandEntry> & out)
+{
+    ServerClusterInterface * serverCluster = GetServerClusterInterface(path);
+    VerifyOrReturnError(serverCluster != nullptr, CHIP_ERROR_NOT_FOUND);
+    return serverCluster->AcceptedCommands(path, out);
+}
+
+CHIP_ERROR CodeDrivenDataModelProvider::Attributes(const ConcreteClusterPath & path,
+                                                   ReadOnlyBufferBuilder<DataModel::AttributeEntry> & out)
+{
+    ServerClusterInterface * serverCluster = GetServerClusterInterface(path);
+    VerifyOrReturnError(serverCluster != nullptr, CHIP_ERROR_NOT_FOUND);
+    return serverCluster->Attributes(path, out);
+}
+
+CHIP_ERROR CodeDrivenDataModelProvider::EventInfo(const ConcreteEventPath & path, DataModel::EventEntry & eventInfo)
+{
+    ServerClusterInterface * serverCluster = GetServerClusterInterface(path);
+    VerifyOrReturnError(serverCluster != nullptr, CHIP_ERROR_NOT_FOUND);
+    return serverCluster->EventInfo(path, eventInfo);
+}
+
+void CodeDrivenDataModelProvider::Temporary_ReportAttributeChanged(const AttributePathParams & path)
+{
+    if (mContext.dataModelChangeListener == nullptr)
+    {
+        ChipLogError(DataManagement, "Temporary_ReportAttributeChanged called before provider has been started.");
+        return;
+    }
+    mContext.dataModelChangeListener->MarkDirty(path);
+}
+
+CHIP_ERROR CodeDrivenDataModelProvider::AddEndpoint(EndpointInterfaceRegistration & registration)
+{
+    VerifyOrReturnError(registration.endpointInterface != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+
+    if (registration.endpointEntry.id == kInvalidEndpointId)
+    {
+        registration.endpointEntry.id = mNextAvailableEndpointId;
+    }
+    else if (registration.endpointEntry.id < mNextAvailableEndpointId)
+    {
+        // If the endpoint ID is already in use, return an error.
+        return CHIP_ERROR_DUPLICATE_KEY_ID;
+    }
+
+    ReturnErrorOnFailure(mEndpointInterfaceRegistry.Register(registration));
+
+    // Update the next available endpoint ID
+    mNextAvailableEndpointId = registration.endpointEntry.id + 1;
```

Comment:
I somehow feel uneasy about this ... this means I can go back in time and make using kInvalidEndpointId not work anymore. It is indeed invalid to re-register and endpoint, however how about:

```suggestion
    mNextAvailableEndpointId = std::max(mNextAvailableEndpointId, registration.endpointEntry.id + 1);
```

Reply:
Sure, I've made the suggested change.

But since we have the check `if (registration.endpointEntry.id < mNextAvailableEndpointId)` prior, std::max(mNextAvailableEndpointId, registration.endpointEntry.id + 1);` is always going to resolve to `registration.endpointEntry.id + 1` anyways but I guess this adds another line of defense. So I made the change.

Thanks!